[quant] Input Weight Equalization - prepare modifications

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59474 [quant] Input Weight Equalization - prepare modifications**

Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28968773](https://our.internmc.facebook.com/intern/diff/D28968773)

angelayi committed:
[quant] Input Weight Equalization - prepare modifications

Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -92,6 +97,8 @@ def calculate_qparams(self):
 
         return scale_input, zero_point_input
 
+    with_args = classmethod(_with_args)
```
should input observer and weight observer inherit from ObserverBase? https://github.com/pytorch/pytorch/blob/master/torch/quantization/observer.py#L47

angelayi replied:
Sure! I can change that.

jerryzh168 commented on torch/quantization/fx/prepare.py:
```diff
@@ -458,7 +460,10 @@ def maybe_insert_output_observer_for_node(
+
             observer = act_post_process_ctr()
+            if isinstance(observer, _InputEqualizationObserver):
+                observer = observer.output_obs
```
I feel output_obs needs to be a constructor as well (see comment in previous PR)

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -110,3 +124,48 @@ def test_input_weight_observer(self, input_qdtype, input_qscheme, weight_qdtype,
         self.assertTrue(torch.allclose(weight_qparams[1], torch.tensor(
             ref_zero_points, dtype=weight_qparams[1].dtype), atol=1))
+
+    def test_input_weight_equalization_prepare(self):
```
Can we also add a couple different models to this test
1. test nn.functional.linear
2. Linear followed by fp32 op
2. Linear followed by another quantized op (non-linear)

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/_equalize.py:
```diff
@@ -1,11 +1,13 @@
 
 
-class _InputEqualizationObserver(nn.Module):
+class _InputEqualizationObserver(ObserverBase):
```
I'm not super familiar with the equalization code, but are we sure this is the right abstraction to have this inherit from `ObserverBase`?  It seems unexpected that an instance of `ObserverBase` would then have children that are also instances of `ObserverBase`.  Thoughts?

angelayi replied:
I originally had it as an `nn.Module` but with the new changes, in order to create a default QConfig, I needed a `with_args` function which is located in `ObserverBase`. @jerryzh168 suggested to have the class inherit `ObserverBase` which made sense since this class functions as an Observer. Conversely, another solution which I had before could be to just import `_with_args()` and create a `with_args` function separately in `_InputEqualizationObserver` and `_WeightEqualizationObserver`.

vkuzo replied:
IMO might be worth considering not having a class have children of the same base class, it might be hard for code readers to understand.  I'd vote for sticking with `nn.Module` as base class.

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -110,3 +124,142 @@ def test_input_weight_observer(self, input_qdtype, input_qscheme, weight_qdtype,
+        prepared = prepare_fx(m, qconfig_dict)
+
+        node_occurrence = {
+            ns.call_module(_InputEqualizationObserver): 1,
+            ns.call_module(_WeightEqualizationObserver): 2,
+            ns.call_module(MinMaxObserver): 2,
```
For multiple Linear layers I would expect the inputEqualizationObserver to be inserted at the input of each of them to calculate the scale factors. We probably need to check the next observer type (Equalization/nonEqualization) before we always default to MinMaxObserver.

angelayi replied:
I addressed this in prepare.py where before inserting the default observer, we check that the next node's observer is not an _InputEqualizationObserver. Does this change seem ok?

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -110,3 +124,142 @@ def test_input_weight_observer(self, input_qdtype, input_qscheme, weight_qdtype,
+
+    def test_input_weight_equalization_prepare_functional_linear_op(self):
+        """ Tests that observers are inserted correctly for two functional
+        linear layers with a fp32 operation in between
```
The mul operator in between is still quantized (you can call convert_fx to confirm this) because you specified the default qconfig for all operators. To simulate an fp32 operator you could add, say a conv in between two linear layers and set the qconfig dict to default to None for all other ops `qconfig_dict = { "" : None, ...}`

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -110,3 +124,142 @@ def test_input_weight_observer(self, input_qdtype, input_qscheme, weight_qdtype,
             ref_scales, dtype=weight_qparams[0].dtype), atol=0.0001))
         self.assertTrue(torch.allclose(weight_qparams[1], torch.tensor(
             ref_zero_points, dtype=weight_qparams[1].dtype), atol=1))
+
```
General comment: For all these tests seems like just the model type and node_occurance is changing. Thoughts on consolidating it into one/two tests and looping over the different types?

angelayi replied:
Ah, that would be a more concise way -- would this be structured as looping over a list of (model type, node_occurance) tuples?

supriyar replied:
yes, I think there are a few tests in test_quantize_fx that are already doing this.

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo reviewed (changes requested):
looks good, requesting changes for not inheriting from `ObserverBase` if we have observer children.  One thing to consider for precedent is the `FakeQuantize` class, which also does similar things to observers and has observers as children, but does not inherit from `ObserverBase`.

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28968773).

vkuzo commented on torch/quantization/fx/_equalize.py:
```diff
@@ -40,18 +43,16 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
-                                             quant_min=quant_min,
-                                             quant_max=quant_max,
-                                             factory_kwargs=factory_kwargs)
+            self.output_obs = MinMaxObserver
```
nit: there could be some confusion if `input_obs` refers to an observer instance but `output_obs` refers to an observer class type.  Some strategies to improve this could be to name the variables differently (i.e. `output_obs_cls`), specify the types, store the instance instead of the class, etc.

angelayi replied:
That makes sense; I will fix that!

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28968773](https://our.internmc.facebook.com/intern/diff/D28968773)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28968773](https://our.internmc.facebook.com/intern/diff/D28968773)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/utils.py:
```diff
@@ -469,3 +469,15 @@ def is_get_tensor_info_node(node: Node) -> bool:
         node.op == "call_function" and node.target == getattr and node.args[1] == "shape"  # type: ignore[assignment]
     return result
+
+def get_next_node(model: GraphModule, node: Node) -> Optional[Node]:
```
Can we just do `node.next()`?

angelayi replied:
I tried using `node.next` but it didn't give the wanted values. For example, for the graph with 2 consecutive linear layers, if node is `linear`, `get_next_node()` would return `linear_1` which we can use to search in `matches`, but `node.next` would return `linear2_w` which doesn't exist in `matches`.

supriyar replied:
Would it work if we kept calling node.next in a loop until we hit a `('call_module', 'call_method', 'call_function', 'output')` type?

jerryzh168 commented on torch/quantization/fx/prepare.py:
```diff
@@ -458,7 +461,19 @@ def maybe_insert_output_observer_for_node(
             observer = act_post_process_ctr()
+            if isinstance(observer, _InputEqualizationObserver):
+                # Find the next node in the graph
+                next_node = get_next_node(model, node)
```
is this trying to find the observer for the node? I think we should use: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantization_patterns.py#L79

angelayi replied:
I wanted to get the observer of the next node. I wasn't sure if the maybe_get_last_node_only() function would get the next node's observer.

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -458,7 +461,19 @@ def maybe_insert_output_observer_for_node(
                         act_post_process_ctr)
+
             observer = act_post_process_ctr()
+            if isinstance(observer, _InputEqualizationObserver):
```
we should only insert output observers inside of the `maybe_insert_output_observer_for_node`.  If I understand this PR correctly, a better place for inserting the equalization input observer object would be somewhere in `maybe_insert_input_observers_for_node`.
